californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -396,11 +398,12 @@ public void completed(final Exchange exchange) {
 
 		@Override
 		public void contextEstablished(final Exchange exchange) {
-
-			if (exchange.getRequest() != null) {
-				observationStore.setContext(exchange.getRequest().getToken(), exchange.getCorrelationContext());
+			Request currentRequest = exchange.getCurrentRequest(); 
+			Request request = exchange.getRequest(); 
+			if (request == currentRequest && request != null && request.isObserve()) {
+				observationStore.setContext(request.getToken(), exchange.getCorrelationContext());
```

Comment:
`correlationContext != ctx`

I guess, this will always fire :-). (currently it's implemented as always a new object).

So, this triggers the question, about the natures of correlation contexts. 
Should we replace the "==" by "some equals"?
Should the connector reuse the CorrelationContext object?

By the way, my last comment on issue #311 will still be valid even with this improvement.

https://github.com/eclipse/californium/issues/311#issuecomment-305957631